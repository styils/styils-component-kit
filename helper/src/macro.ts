import { createMacro } from 'babel-plugin-macros'
import * as t from '@babel/types'
import { addNamed } from '@babel/helper-module-imports'
import type { NodePath } from '@babel/traverse'
import type { MParams } from './types'
import { component } from './component'
import { state } from './state'
import { proper } from './proper'
import { useMount } from './useMount'
import { useMemo } from './useMemo'
import { useWatchEffect } from './useWatchEffect'

/**
 * Avoid importing multiple identical methods
 */
const hookCacheId = new Map()

export default createMacro(({ references, state: babelState }) => {
  /**
   * used to get all props states
   * Used to get all props state, auto-populate react deps
   */
  const variableMaps = new Set([])
  const {
    component: Icomponent = [],
    proper: Iproper = [],
    state: Istate = [],
    ...macros
  } = references

  const createAddImportPath = (path: NodePath) => (name: string, source: string) => {
    const cacheCode = babelState.file.code + name

    const nameId = hookCacheId.get(cacheCode) ?? addNamed(path, name, source)
    hookCacheId.set(cacheCode, nameId)

    return nameId
  }

  // component always stays first
  Icomponent.forEach((path) => {
    component(path, (babelState.opts as { frame: string }).frame, createAddImportPath(path))
  })

  Iproper.forEach((path) => {
    proper(path, { addImportName: createAddImportPath(path), ...babelState }, variableMaps)
  })

  Istate.forEach((path) => {
    state(path, { addImportName: createAddImportPath(path), ...babelState }, variableMaps)
  })

  const macroHookMethods = [useMount, useMemo, useWatchEffect]

  macroHookMethods.forEach((macro) => {
    macros[macro.name]?.forEach((path) => {
      if (!t.isCallExpression(path.parentPath.node)) {
        // Macros must be called directly, otherwise the correct call cannot be traced
        throw new Error('All macros must be called')
      }

      const currentCallExpression = path.findParent((p) =>
        t.isCallExpression(p)
      ) as MParams['currentCallExpression']

      macro(
        path,
        {
          currentCallExpression,
          ...babelState, // `import .. form ..` helper method
          addImportName: createAddImportPath(path)
        },
        variableMaps
      )
    })
  })
})
