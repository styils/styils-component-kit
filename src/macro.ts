import { createMacro } from 'babel-plugin-macros'
import * as t from '@babel/types'
import { addNamed } from '@babel/helper-module-imports'
import type { NodePath } from '@babel/traverse'
import type { MParams } from './types'
import { component } from './component'
import { state } from './state'
import { proper } from './proper'
import { ref } from './ref'
import { useMount } from './useMount'
import { useMemo } from './useMemo'
import { useWatchEffect } from './useWatchEffect'

/**
 * Avoid importing multiple identical methods
 */
const hookCacheId = new Map()

export default createMacro(({ references, state: babelState }) => {
  /**
   * used to get all props states
   * Used to get all props state, auto-populate react deps
   */
  const variableMaps = new Set([])
  const { component: Icomponent = [], proper: Iproper = [], ...macros } = references

  const createAddImportPath = (path: NodePath) => (name: string, source: string) => {
    const cacheCode = babelState.file.code + name

    const nameId = hookCacheId.get(cacheCode) ?? addNamed(path, name, source)
    hookCacheId.set(cacheCode, nameId)

    return nameId
  }

  // component always stays first
  Icomponent.forEach((path) => {
    component(path, (babelState.opts as { frame: string }).frame, createAddImportPath(path))
  })

  Iproper.forEach((path) => {
    proper(path, { addImportName: createAddImportPath(path), ...babelState }, variableMaps)
  })

  const macroStateMethods = [state, ref]

  macroStateMethods.forEach((macro) => {
    macros[macro.name]?.forEach((path) => {
      if (!t.isCallExpression(path.parentPath.node)) {
        // Macros must be called directly, otherwise the correct call cannot be traced
        throw new Error('All macros must be called')
      }

      const currentVariableDeclaration = path.findParent((p) =>
        t.isVariableDeclaration(p)
      ) as MParams['currentVariableDeclaration']

      const currentCallExpression = path.findParent((p) =>
        t.isCallExpression(p)
      ) as MParams['currentCallExpression']

      const currentFunction = path.getFunctionParent() as MParams['currentFunction']
      const currentVariableDeclarator = path.parentPath.parentPath.node

      const identifiers: NodePath<t.Identifier>[] = []
      const setIdentifiers: NodePath<t.CallExpression>[] = []

      if (t.isVariableDeclarator(currentVariableDeclarator)) {
        if (
          t.isIdentifier(currentVariableDeclarator.id) ||
          t.isArrayPattern(currentVariableDeclarator.id)
        ) {
          let identNodeName: { name: string; start: number } | undefined
          let setIdentNode: { name: string; start: number } | undefined

          if (t.isIdentifier(currentVariableDeclarator.id)) {
            identNodeName = {
              start: currentVariableDeclarator.id.start,
              name: currentVariableDeclarator.id.name
            }
          }

          if (
            t.isArrayPattern(currentVariableDeclarator.id) &&
            t.isIdentifier(currentVariableDeclarator.id.elements[0])
          ) {
            const [variable, setVariable] = currentVariableDeclarator.id.elements
            if (t.isIdentifier(variable)) {
              identNodeName = {
                start: variable.start,
                name: variable.name
              }
            }

            if (t.isIdentifier(setVariable)) {
              setIdentNode = {
                start: setVariable.start,
                name: setVariable.name
              }
            }
          }

          currentFunction.traverse({
            Identifier(IPath) {
              if (
                setIdentNode &&
                IPath.node.name === setIdentNode.name &&
                // exclude initialization nodes
                setIdentNode.start !== IPath.node.start &&
                // and new replacement nodes, new node without start
                IPath.node.start &&
                currentFunction.scope.hasOwnBinding(IPath.node.name) &&
                t.isCallExpression(IPath.parentPath)
              ) {
                // handling set methods
                setIdentifiers.push(IPath.parentPath as NodePath<t.CallExpression>)
              }

              if (
                identNodeName &&
                IPath.node.name === identNodeName.name &&
                // exclude initialization nodes
                identNodeName.start !== IPath.node.start &&
                // and new replacement nodes, new node without start
                IPath.node.start &&
                !t.isVariableDeclaration(IPath.parentPath.node) &&
                currentFunction.scope.hasOwnBinding(IPath.node.name)
              ) {
                // If it is an assignment expression
                // the current scope is saved
                // the assignment always precedes the use
                let scoper = IPath.scope

                while (scoper) {
                  if (scoper.hasOwnBinding(IPath.node.name)) {
                    if (scoper.block.start === path.scope.block.start) {
                      break
                    } else {
                      return
                    }
                  }

                  scoper = scoper.parent
                }

                const isRefJsxValue = IPath.findParent(
                  (item) =>
                    t.isJSXAttribute(item.node) &&
                    t.isJSXIdentifier(item.node.name) &&
                    item.node.name.name === 'ref'
                )

                if (!isRefJsxValue) {
                  identifiers.push(IPath)
                }
              }
            }
          })
        }
      }

      macro(
        path,
        {
          // `import .. form ..` helper method
          addImportName: createAddImportPath(path),
          currentFunction,
          currentCallExpression,
          currentVariableDeclaration,
          currentVariableDeclarator,
          identifiers,
          setIdentifiers,

          ...babelState
        },
        variableMaps
      )
    })
  })

  const macroHookMethods = [useMount, useMemo, useWatchEffect]

  macroHookMethods.forEach((macro) => {
    macros[macro.name]?.forEach((path) => {
      if (!t.isCallExpression(path.parentPath.node)) {
        // Macros must be called directly, otherwise the correct call cannot be traced
        throw new Error('All macros must be called')
      }

      const currentCallExpression = path.findParent((p) =>
        t.isCallExpression(p)
      ) as MParams['currentCallExpression']

      macro(
        path,
        {
          currentCallExpression,
          ...babelState, // `import .. form ..` helper method
          addImportName: createAddImportPath(path)
        },
        variableMaps
      )
    })
  })
})
