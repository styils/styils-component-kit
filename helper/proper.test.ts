import { getCode } from 'test'

describe('use props', () => {
  it('base react', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo, bar} = proper(props)

      return (
        <button>
          Clicked {foo} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'react' }
    )

    expect(code).toMatchSnapshot()
  })

  it('react merge props', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo, bar} = proper(props, {foo:1})

      return (
        <button>
          Clicked {foo} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'react' }
    )

    expect(code).toMatchSnapshot()
  })

  it('base vue', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo, bar} = proper(props)

      function onclick(){
        let foo = 0

        if (a == 1){
          if (b == 2){
            foo++
          }
        }
      }

      return (
        <button>
          Clicked {foo} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'vue' }
    )

    expect(code).toMatchSnapshot()
  })

  it('vue merge props', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo, bar} = proper(props, {foo:1})

      return (
        <button>
          Clicked {foo} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'vue' }
    )

    expect(code).toMatchSnapshot()
  })

  it('vue scope proper', () => {
    const code = getCode(
      `
      import { proper } from 'macro'

      export default (props) => {
        const {foo:foo1, bar, ...rest} = proper(props, {foo:'1'})

        function onclick(){
          let foo1 = 0

          if (a == 1){
            if (b == 2){
              foo1++
            }
          }
        }

        function onclick1(){
          foo1++

          if (a == 1){
            let foo1 = 0
            if (b == 2){
              foo1++
            }
          }
        }

        return (
          <button {...rest}>
            Clicked {foo1} {bar === 1 ? 'time' : 'times'}
          </button>
        )
      }
    `,
      { frame: 'vue' }
    )

    expect(code).toMatchSnapshot()
  })

  it('base solid', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo:foo1, bar, ...rest} = proper(props)

      return (
        <button {...rest}>
          Clicked {foo1} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'solid' }
    )

    expect(code).toMatchSnapshot()
  })

  it('solid merge props', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo:foo1, bar, ...rest} = proper(props, {foo:'1'})

      return (
        <button {...rest}>
          Clicked {foo1} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'solid' }
    )

    expect(code).toMatchSnapshot()
  })

  it('solid scope proper', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo:foo1, bar, ...rest} = proper(props, {foo:'1'})

      function onclick(){
        let foo1 = 0

        if (a == 1){
          if (b == 2){
            foo1++
          }
        }
      }

      function onclick1(){
        foo1++

        if (a == 1){
          let foo1 = 0
          if (b == 2){
            foo1++
          }
        }
      }

      return (
        <button {...rest}>
          Clicked {foo1} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'solid' }
    )

    expect(code).toMatchSnapshot()
  })
})
