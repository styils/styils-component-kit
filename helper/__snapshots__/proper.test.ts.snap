// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use props base react 1`] = `
"export default (props) => {
  const { foo, bar } = props;
  return (
    <button>
      Clicked {foo} {bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props base solid 1`] = `
"import { splitProps as _splitProps } from "solid-js";
export default (props) => {
  const [_local_, rest] = _splitProps(props, ["foo", "bar"]);

  return (
    <button {...rest}>
      Clicked {_local_.foo} {_local_.bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props base vue 1`] = `
"import { toRefs as _toRefs } from "vue";
export default (props) => {
  const { foo, bar } = _toRefs(props);

  function onclick() {
    let foo = 0;

    if (a == 1) {
      if (b == 2) {
        foo++;
      }
    }
  }

  return (
    <button>
      Clicked {foo.value} {bar.value === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props react merge props 1`] = `
"export default (props) => {
  const { foo, bar } = {
    ...{
      foo: 1,
    },
    ...props,
  };
  return (
    <button>
      Clicked {foo} {bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props solid merge props 1`] = `
"import { mergeProps as _mergeProps } from "solid-js";
import { splitProps as _splitProps } from "solid-js";
export default (props) => {
  const [_local_, rest] = _splitProps(
    _mergeProps(
      {
        foo: "1",
      },
      props
    ),
    ["foo", "bar"]
  );

  return (
    <button {...rest}>
      Clicked {_local_.foo} {_local_.bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props solid scope proper 1`] = `
"import { mergeProps as _mergeProps } from "solid-js";
import { splitProps as _splitProps } from "solid-js";
export default (props) => {
  const [_local_, rest] = _splitProps(
    _mergeProps(
      {
        foo: "1",
      },
      props
    ),
    ["foo", "bar"]
  );

  function onclick() {
    let foo1 = 0;

    if (a == 1) {
      if (b == 2) {
        foo1++;
      }
    }
  }

  function onclick1() {
    _local_.foo++;

    if (a == 1) {
      let foo1 = 0;

      if (b == 2) {
        foo1++;
      }
    }
  }

  return (
    <button {...rest}>
      Clicked {_local_.foo} {_local_.bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props vue merge props 1`] = `
"import { reactive as _reactive } from "vue";
import { readonly as _readonly } from "vue";
import { toRefs as _toRefs } from "vue";
export default (props) => {
  const { foo, bar } = _toRefs(
    _readonly(
      _reactive({
        ...{
          foo: 1,
        },
        ...props,
      })
    )
  );

  return (
    <button>
      Clicked {foo.value} {bar.value === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props vue scope proper 1`] = `
"import { reactive as _reactive } from "vue";
import { readonly as _readonly } from "vue";
import { toRefs as _toRefs } from "vue";
export default (props) => {
  const { foo: foo1, bar } = _toRefs(
    _readonly(
      _reactive({
        ...{
          foo: "1",
        },
        ...props,
      })
    )
  );

  function onclick() {
    let foo1 = 0;

    if (a == 1) {
      if (b == 2) {
        foo1++;
      }
    }
  }

  function onclick1() {
    foo1.value++;

    if (a == 1) {
      let foo1 = 0;

      if (b == 2) {
        foo1++;
      }
    }
  }

  return (
    <button {...rest}>
      Clicked {foo1.value} {bar.value === 1 ? "time" : "times"}
    </button>
  );
};
"
`;
